### DISCUSSION TITLE
Error with installation

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#20618] [Q&A Getting Started] [answered]
Hello,
currently i'm updating the version of MOOSE I am currently using.
I'm following HPC install steps, and loading 
cmake/3.20.6
gcc/10.2.0
openmpi/4.0.5
this compiles PETSc and libmesh just fine, but when i try running the tests, it doesn't detect the python installation. for that I need to load anaconda 

however, whenever I try loading anaconda, installing PETSc fails with the error that there are no C compiler found. 

any suggestions?

**Comment by GiudGiud** [comment, marked as answer]
Hello

If loading your conda environment comes with a set of compilers, a different MPI library, etc, then this will set it up for failure, as there will be conflicts between the modules you loaded and the conda packages. 
Is there a python module you can load instead? Or a clean anaconda environment?

Guillaume

**Reply by KhaledNabilSharafeldin** [reply]
I did ask the cluster team if there's a separate python module, but unfortunately no, that isn't the case.

However, I tried loading up anaconda then creating a bare down environment, then activated it and tried to compile PETSc. I get the same error.  ``Could not locate a functional C compiler``
used this to create the env ``conda create --name tempenv python=3.6``

all $CC and other env vars are set correctly.

**Reply by KhaledNabilSharafeldin** [reply]
i'm wondering if these installation steps need to compile both PETSc and libmesh afterwards or not

https://mooseframework.inl.gov/getting_started/installation/conda.html

**Reply by GiudGiud** [reply]
> i'm wondering if these installation steps need to compile both PETSc and libmesh afterwards or not
> 
> https://mooseframework.inl.gov/getting_started/installation/conda.html

These installation steps are not for clusters. They are for local machines. You cannot mix these instructions at all. 

> I did ask the cluster team if there's a separate python module, but unfortunately no, that isn't the case.
> 
> However, I tried loading up anaconda then creating a bare down environment, then activated it and tried to compile PETSc. I get the same error. `Could not locate a functional C compiler` used this to create the env `conda create --name tempenv python=3.6`
> 
> all $CC and other env vars are set correctly.

What are CC and other env vars?
If it cant find a C compiler, there could be an issue.

what do these return ?
` echo $CC $CXX $FC `
and 
`which mpirun `
and
`module list`

and paste the output of `env` as well please, just to check

**Reply by KhaledNabilSharafeldin** [reply]
~ $ echo $CC $CXX $FC
mpicc mpiCC mpif90

:~ $ which mpirun
/apps/cent7/xalt/bin/mpirun

:~ $ module list

Currently Loaded Modules:
  1) xalt/1.1.2 (S)   3) gmp/6.1.2    5) mpc/1.1.0    7) zlib/1.2.11     9) cmake/3.20.6
  2) rcac             4) mpfr/3.1.6   6) gcc/10.2.0   8) openmpi/4.0.5

  Where:
   S:  Module is Sticky, requires --force to unload or purge


conda env dir: ~/.conda/envs/cent7/2020.02-py37/tempenv

**Reply by GiudGiud** [reply]
Never seen this mpiCC , what does this return: `mpiCC --version`
I'm not familiar with mpc? Is that an mpi distribution too? Do you need this?

Can you try with as few modules as possible? And conda does not need to be loaded when installing. 
Python only helps for some of the tests.

**Reply by KhaledNabilSharafeldin** [reply]
:~ $ mpiCC --version
g++ (Spack GCC) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Reply by KhaledNabilSharafeldin** [reply]
the reason I suspected a faulty installation is that when I compile my project, one of the errors i get is that 
``exII`` is not a namespace. which i use in one of the source files to generate a mesh.

not really sure what the problem is since this namespace is defined in "libmesh/exodusII_io_helper.h" source file and i do include it. This is the first time I get this kind of error.

not to mention a bunch of errors due to old inputparameters syntax, which does not go away even after updating the syntax for it (no forward declaration, no template, etc..., basically following the example in the link provided by the error https://mooseframework.inl.gov/newsletter/2021_11.html#legacy-input-parameter-deprecation)

**Reply by GiudGiud** [reply]
Can you paste one of those inputParam errors here?
And the corresponding code and header for which you get the error.
Links to pages on a github repo would work too obviously

This is all unrelated to a `missing C compiler` error.

**Reply by KhaledNabilSharafeldin** [reply]
Terminal output
```
projects/CDD-MOOSE/src/main.C:24:17: error: 'exII' is not a namespace-name
   24 | using namespace exII;
      |                 ^~~~
```

code snippet:
```
#include "cddAppTestApp.h"
#include "MooseInit.h"
#include "Moose.h"
#include "MooseApp.h"
#include "AppFactory.h"

#include "libmesh/exodusII_io_helper.h"
// #include "libmesh/exodusII.h"
double L_MESH;
double X_MIN;
double X_MAX;
double Y_MIN;
double Y_MAX;
double Z_MIN;
double Z_MAX;
using namespace exII;
#include "GlobalVars.h"
#include "libmesh/boundary_info.h"
#include "libmesh/enum_elem_type.h"
#include "libmesh/elem.h"
```




Terminal error
<pre>In file included from <b>/home/projects/moose/framework/build/header_symlinks/MooseADWrapper.h:12</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/MooseUtils.h:17</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/InputParameters.h:13</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/DistributionInterface.h:12</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/UserObject.h:13</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/ElementUserObject.h:13</b>,
                 from <b>/home/projects/moose/framework/build/header_symlinks/ElementPostprocessor.h:12</b>,
                 from <b>/home/projects/CDD-MOOSE/build/header_symlinks/ComputeMeanPlasicDistortion.h:11</b>,
                 from <b>/home/projects/CDD-MOOSE/src/postprocessors/ComputeMeanPlasicDistortion.C:10</b>,
                 from <b>/home/projects/CDD-MOOSE/build/unity_src/postprocessors_Unity.C:2</b>:
<b>/home/projects/moose/framework/build/header_symlinks/MooseTypes.h:466:17:</b> <font color="#CC0000"><b>error: </b></font>static assertion failed: defineLegacyParams is no longer supported as legacy input parameter construction is no longer supported; see mooseframework.org/newsletter/2021_11.html#legacy-input-parameter-deprecation for more information
  466 |   static_assert(<font color="#CC0000"><b>false</b></font>,                                                                             \
      |                 <font color="#CC0000"><b>^~~~~</b></font>
<b>/home/projects/moose/framework/build/header_symlinks/MooseTypes.h:466:17:</b> <font color="#06989A"><b>note: </b></font>in definition of macro &apos;<b>defineLegacyParams</b>&apos;
  466 |   static_assert(<font color="#06989A"><b>false</b></font>,                                                                             \
      |                 <font color="#06989A"><b>^~~~~</b></font>
</pre>


source:
```
#include "ComputeMeanPlasicDistortion.h"

#include <algorithm>
#include <limits>

registerMooseObject("cddAppApp", ComputeMeanPlasicDistortion);

InputParameters
ComputeMeanPlasicDistortion::validParams()
{
  // Define the parameters
  InputParameters params = ElementPostprocessor::validParams();
  params.addRequiredCoupledVar("rho_vars","This is the collection of density variables for all slip systems");
  params.addRequiredCoupledVar("z_vars","This is the collection of density variables for all slip systems");
  params.addRequiredCoupledVar("chi_vars","This is the collection of incompatible variables.");
  params.addRequiredCoupledVar("resolved_shear_stress", "The magnitude of the resolved shear stress used to calculate the dislocation velocities ");
  params.addRequiredParam<unsigned int>("i","row id");
  params.addRequiredParam<unsigned int>("j","coloumn id");
  params.addRequiredParam<bool>("relaxed","This specifies whether the initial mean plastic distortion is 0");
  params.addRequiredParam<unsigned int>("number_of_slip_systems"," ");
  params.addParam<bool>("subtract_mean",true,"This is for debugging");
  return params;
}
```



Header:

```
#pragma once

#include "ElementPostprocessor.h"
#include "MicrostructureHelper.h"



/// A postprocessor for collecting the nodal min or max value
class ComputeMeanPlasicDistortion : public ElementPostprocessor
{
public:
  /**
   * Class constructor
   * @param parameters The input parameters
   */

  ComputeMeanPlasicDistortion(const InputParameters & parameters);
  static InputParameters validParams();

  virtual void initialize() override;
  virtual Real getValue() override;
  virtual void execute() override;
  virtual void threadJoin(const UserObject & y) override;
```

**Reply by KhaledNabilSharafeldin** [reply]
I figured out the issue with legacyParamsDefinition, if the original source doesnt have `defineLegacyParam()` it will not prompt an error to that specific file, instead, it'll prompt one of the files within the same folder regardless if it's following the new syntax or not.

still havent figured out what is the issue with `using namespace exII`

**Reply by KhaledNabilSharafeldin** [reply]
Defining the namespace as such solves the problem, thanks!
```
namespace exII {
extern "C" {
#include "libmesh/exodusII.h" // defines MAX_LINE_LENGTH, MAX_STR_LENGTH used later
}
}
using namespace exII;
```

marked as solved

**Reply by KhaledNabilSharafeldin** [reply]
Just a related question to the validity of the installation, whenever I compile the code i get this message at the very end. 

```
/usr/bin/ld: warning: libhwloc.so.5, needed by /home/projects/moose/scripts/../libmesh/installed/lib/libmesh_opt.so, may conflict with libhwloc.so.15
```
is this something i should be worried about? Thanks again!

**Reply by GiudGiud** [reply]
Never had that particular one but had other libs with the same warning and it all went ok. Kudos for backwards compatibility for those developers

**Reply by GiudGiud** [reply]
I'm glad you figured those two out. It's unfortunate for the defineLegacyParam issue. This is almost completely disappeared now so we should never run into this again
